core/src/main/scala/neotypes/Async.scala

```diff
@@ -64,13 +73,26 @@ object Async {
           }
         }
 
+      override final def makeLock: Future[Lock] =
+        Future.successful {
+          val q = new ArrayBlockingQueue[Unit](1)
+
+          new Lock {
+            override final def acquire: Future[Unit] =
+              Future(blocking(q.put(())))
+
+            override final def release: Future[Unit] =
+              Future(blocking(q.take()))
+          }
+        }
+
       override final def map[A, B](fa: Future[A])(f: A => B): Future[B] =
         fa.map(f)
 
       override final def recoverWith[A, B >: A](fa: Future[A])(f: PartialFunction[Throwable, Future[B]]): Future[B] =
         fa.recoverWith(f)
 
-      override final def resource[A](input: => A)(close: A => Future[Unit]): A =
-        input
+      override final def resource[A](input: Future[A])(close: A => Future[Unit]): A =
+        Await.result(input, Duration.Inf)
```

Comment:
Maybe we can make duration configurable somehow? I guess if there is a logical error in the user code that prevents the lock from releasing, user would expect it to time out eventually.

Reply:
There are only two uses of `resource` in all the codebase.
The creation of a **Driver** and the creation of a **Session**, neither of them are waiting for a look and both should be already complete when calling this, so I just used `Await` + `Duration.Inf` to get the plain value from the futures.

Reply:
Gotcha! I hadn't realized that

Reply:
Could also
```java
input.value.get.get
```

Reply:
@geoffjohn11 it should be there, but it could have a couple of milliseconds of delay _(due to the threading allocation of future)_.
Also, all those three methods are unsafe; so I prefer to use `Await` to ensure the value is correctly retrieved.

Now, if we all agree we can use something like `1.second` or `10.seconds` to ensure this doesn't block forever.

Reply:
1.second or 10.seconds sounds good to me